File: browser-extension/src/evageneClient.ts
/**
 * Minimal Evagene REST client used by the background service worker.
 *
 * Owns no state other than its constructor arguments. The caller supplies
 * a fetch implementation so the tests can stand in a fake one without
 * touching global state.
 */

import type { PedigreeSummary } from './messaging.js';

export type FetchLike = (input: string, init?: RequestInit) => Promise<Response>;

export interface EvageneClientOptions {
  readonly baseUrl: string;
  readonly apiKey: string;
  readonly fetch: FetchLike;
}

export class EvageneError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'EvageneError';
  }
}

const HTTP_OK_LOWER = 200;
const HTTP_OK_UPPER = 300;

export class EvageneClient {
  private readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly fetchFn: FetchLike;

  constructor(options: EvageneClientOptions) {
    this.baseUrl = options.baseUrl.replace(/\/$/, '');
    this.apiKey = options.apiKey;
    this.fetchFn = options.fetch;
  }

  async getPedigreeSummary(pedigreeId: string): Promise<PedigreeSummary> {
    const url = `${this.baseUrl}/api/pedigrees/${pedigreeId}/summary`;
    const response = await this.fetchFn(url, { headers: this.headers() });
    if (response.status < HTTP_OK_LOWER || response.status >= HTTP_OK_UPPER) {
      throw new EvageneError(
        `Evagene API returned HTTP ${response.status.toString()} for ${url}`,
      );
    }
    const payload: unknown = await response.json();
    return toSummary(payload, `${this.baseUrl}/pedigrees/${pedigreeId}`);
  }

  private headers(): Record<string, string> {
    return { 'X-API-Key': this.apiKey, Accept: 'application/json' };
  }
}

function toSummary(payload: unknown, viewUrl: string): PedigreeSummary {
  const root = asObject(payload, 'response');
  return {
    pedigreeId: requireString(root, 'pedigree_id'),
    name: requireString(root, 'name'),
    probandName: readProbandName(root.proband),
    diseases: Object.keys(asObject(root.diseases_in_family, 'diseases_in_family')),
    viewUrl,
  };
}

function readProbandName(value: unknown): string | null {
  if (value === null || value === undefined) return null;
  const proband = asObject(value, 'proband');
  const name = proband.name;
  return typeof name === 'string' ? name : null;
}

function asObject(value: unknown, label: string): Record<string, unknown> {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new EvageneError(`${label} is not an object in API response`);
  }
  return value as Record<string, unknown>;
}

function requireString(container: Record<string, unknown>, key: string): string {
  const value = container[key];
  if (typeof value !== 'string') {
    throw new EvageneError(`field '${key}' is missing or not a string`);
  }
  return value;
}

File: pedigree-diff/python/src/pedigree_diff/evagene_client.py
"""Client for ``GET /api/pedigrees/{id}`` — returns the raw PedigreeDetail dict."""

from __future__ import annotations

from typing import Any

from .http_gateway import HttpGateway

_HTTP_OK_LOWER = 200
_HTTP_OK_UPPER = 300


class ApiError(RuntimeError):
    """Raised when the Evagene API is unreachable or returns a non-2xx response."""


class EvageneClient:
    def __init__(self, *, base_url: str, api_key: str, http: HttpGateway) -> None:
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._http = http

    def get_pedigree_detail(self, pedigree_id: str) -> dict[str, Any]:
        url = f"{self._base_url}/api/pedigrees/{pedigree_id}"
        response = self._http.get_json(url, headers=self._headers())
        if not _HTTP_OK_LOWER <= response.status_code < _HTTP_OK_UPPER:
            raise ApiError(f"Evagene API returned HTTP {response.status_code} for {url}")
        payload = response.json()
        if not isinstance(payload, dict):
            raise ApiError(f"Evagene API returned non-object JSON: {type(payload).__name__}")
        return payload

    def _headers(self) -> dict[str, str]:
        return {
            "X-API-Key": self._api_key,
            "Accept": "application/json",
        }
